File: src/nimjl.nim
import os
import arraymancer
import strutils
import strformat

# Const julia path
const csrc_nimjl = "csrc/nimjl.c"
const juliaPath = getEnv("JULIA_PATH")
const juliaIncludesPath = juliaPath / "include" / "julia"
const juliaLibPath = juliaPath / "lib"
const juliaDepPath = juliaPath / "lib" / "julia"
const juliaHeader = "julia.h"

{.passC: "-fPIC".}
{.passC: " -DJULIA_ENABLE_THREADING=1".}
{.passC: "-I" & juliaIncludesPath.}
{.passL: "-L" & juliaLibPath.}
{.passL: "-Wl,-rpath," & juliaLibPath.}
{.passL: "-L" & juliaDepPath.}
{.passL: "-Wl,-rpath," & juliaDepPath.}
{.passL: "-ljulia".}
{.compile: csrc_nimjl.}

static:
  echo "juliaPath> ", juliaPath
  echo "juliaIncludesPath> ", juliaIncludesPath
  echo "juliaLibPath> ", juliaLibPath
# {.push header: juliaHeader.}

##Types
type nimjl_value *{.importc: "jl_value_t", header: juliaHeader.} = object
type nimjl_array *{.importc: "jl_array_t", header: juliaHeader.} = object
type nimjl_func *{.importc: "jl_function_t", header: juliaHeader.} = object
type nimjl_module *{.importc: "jl_module_t", header: juliaHeader.} = object

var jl_main_module *{.importc: "jl_main_module", header: juliaHeader.}: ptr nimjl_module
var jl_core_module *{.importc: "jl_core_module", header: juliaHeader.}: ptr nimjl_module
var jl_base_module *{.importc: "jl_base_module", header: juliaHeader.}: ptr nimjl_module
var jl_top_module *{.importc: "jl_top_module", header: juliaHeader.}: ptr nimjl_module

## Basic function
proc nimjl_init*() {.cdecl, importc.}
proc nimjl_gc_enable*(toggle: cint) {.cdecl, importc.}
proc nimjl_atexit_hook*(exit_code: cint) {.cdecl, importc.}
proc nimjl_eval_string*(code: cstring): ptr nimjl_value {.cdecl, importc.}

proc nimjl_eval_string*(code: string): ptr nimjl_value =
  result = nimjl_eval_string(code.cstring)

## Box & Unbox
proc nimjl_unbox_float64(value: ptr nimjl_value): float64 {.cdecl, importc.}
proc nimjl_unbox_float32(value: ptr nimjl_value): float32 {.cdecl, importc.}

proc nimjl_unbox_int64(value: ptr nimjl_value): int64 {.cdecl, importc.}
proc nimjl_unbox_int32(value: ptr nimjl_value): int32 {.cdecl, importc.}
proc nimjl_unbox_int16(value: ptr nimjl_value): int16 {.cdecl, importc.}
proc nimjl_unbox_int8(value: ptr nimjl_value): int8 {.cdecl, importc.}

proc nimjl_unbox_uint64(value: ptr nimjl_value): uint64 {.cdecl, importc.}
proc nimjl_unbox_uint32(value: ptr nimjl_value): uint32 {.cdecl, importc.}
proc nimjl_unbox_uint16(value: ptr nimjl_value): uint16 {.cdecl, importc.}
proc nimjl_unbox_uint8(value: ptr nimjl_value): uint8 {.cdecl, importc.}

proc nimjl_box_float64(value: float64): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_float32(value: float32): ptr nimjl_value {.cdecl, importc.}

proc nimjl_box_int64(value: int64): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_int32(value: int32): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_int16(value: int16): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_int8(value: int8): ptr nimjl_value {.cdecl, importc.}

proc nimjl_box_uint64(value: uint64): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_uint32(value: uint32): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_uint16(value: uint16): ptr nimjl_value {.cdecl, importc.}
proc nimjl_box_uint8(value: uint8): ptr nimjl_value {.cdecl, importc.}

proc nimjl_unbox*[T](value: ptr nimjl_value): T =
  when T is int8:
    result = nimjl_unbox_int8(value)
  elif T is int16:
    result = nimjl_unbox_int16(value)
  elif T is int32:
    result = nimjl_unbox_int32(value)
  elif T is int64:
    result = nimjl_unbox_int64(value)
  elif T is uint8:
    result = nimjl_unbox_uint8(value)
  elif T is uint16:
    result = nimjl_unbox_uint16(value)
  elif T is uint32:
    result = nimjl_unbox_uint32(value)
  elif T is uint64:
    result = nimjl_unbox_uint64(value)
  elif T is float32:
    result = nimjl_unbox_float32(value)
  elif T is float64:
    result = nimjl_unbox_float64(value)
  else:
    doAssert(false, "Type not supported")

proc nimjl_box*[T](value: T): ptr nimjl_value =
  when T is int8:
    result = nimjl_box_int8(value)
  elif T is int16:
    result = nimjl_box_int16(value)
  elif T is int32:
    result = nimjl_box_int32(value)
  elif T is int64:
    result = nimjl_box_int64(value)
  elif T is uint8:
    result = nimjl_box_uint8(value)
  elif T is uint16:
    result = nimjl_box_uint16(value)
  elif T is uint32:
    result = nimjl_box_uint32(value)
  elif T is uint64:
    result = nimjl_box_uint64(value)
  elif T is float32:
    result = nimjl_box_float32(value)
  elif T is float64:
    result = nimjl_box_float64(value)
  else:
    doAssert(false, "Type not supported")

##GC Functions

proc nimjl_gc_push1*(a: pointer) {.cdecl, importc.}

proc nimjl_gc_push2*(a: pointer, b: pointer) {.cdecl, importc.}

proc nimjl_gc_push3*(a: pointer, b: pointer, c: pointer) {.cdecl, importc.}

proc nimjl_gc_push4*(a: pointer, b: pointer, c: pointer, d: pointer) {.cdecl, importc.}

proc nimjl_gc_push5*(a: pointer, b: pointer, c: pointer, d: pointer,
        e: pointer) {.cdecl, importc.}

proc nimjl_gc_push6*(a: pointer, b: pointer, c: pointer, d: pointer, e: pointer,
        f: pointer) {.cdecl, importc.}

proc nimjl_gc_pushargs*(a: pointer, n: csize_t) {.cdecl, importc.}

proc nimjl_gc_pop*() {.cdecl, importc.}

proc nimjl_exception_occurred*(): ptr nimjl_value {.cdecl, importc.}

proc nimjl_typeof_str*(v: ptr nimjl_value): cstring {.cdecl, importc.}

proc nimjl_string_ptr*(v: ptr nimjl_value): cstring {.cdecl, importc.}

## Call functions
proc nimjl_get_function*(module: ptr nimjl_module, name: cstring): ptr nimjl_func {.cdecl, importc.}

proc nimjl_call *(function: ptr nimjl_func, values: ptr ptr nimjl_value, nargs: cint): ptr nimjl_value {.cdecl, importc.}

proc nimjl_call0*(function: ptr nimjl_func): ptr nimjl_value {.cdecl, importc.}

proc nimjl_call1*(function: ptr nimjl_func, arg: ptr nimjl_value): ptr nimjl_value {.cdecl, importc.}

proc nimjl_call2*(function: ptr nimjl_func, arg1: ptr nimjl_value, arg2: ptr nimjl_value): ptr nimjl_value {.cdecl, importc.}

proc nimjl_call3*(function: ptr nimjl_func, arg1: ptr nimjl_value, arg2: ptr nimjl_value, arg3: ptr nimjl_value): ptr nimjl_value {.cdecl, importc.}

## Check for nil result
proc nimjl_include_file*(file_name: string): ptr nimjl_value =
  result = nimjl_eval_string(&"include(\"{file_name}\")")

proc nimjl_using_module*(module_name: string): ptr nimjl_value =
  result = nimjl_eval_string(&"using {module_name}")

proc nimjl_get_module*(module_name: string): ptr nimjl_module =
  result = cast[ptr nimjl_module](nimjl_eval_string(module_name))

proc nimjl_exec_func*(module: ptr nimjl_module, func_name: string, va: varargs[ptr nimjl_value]): ptr nimjl_value =
  let f = nimjl_get_function(module, func_name)
  nimjl_gc_push1(f.unsafeAddr)

  if va.len == 0:
    result = nimjl_call0(f)
  elif va.len == 1:
    result = nimjl_call1(f, va[0])
  elif va.len == 2:
    result = nimjl_call2(f, va[0], va[1])
  elif va.len == 3:
    result = nimjl_call3(f, va[0], va[1], va[2])
  else:
    result = nimjl_call(f, unsafeAddr(va[0]), va.len.cint)

  nimjl_gc_pop()

proc nimjl_exec_func*(func_name: string, va: varargs[ptr nimjl_value]): ptr nimjl_value =
  let f = nimjl_get_function(jl_main_module, func_name)
  nimjl_gc_push1(f.unsafeAddr)

  if va.len == 0:
    result = nimjl_call0(f)
  elif va.len == 1:
    result = nimjl_call1(f, va[0])
  elif va.len == 2:
    result = nimjl_call2(f, va[0], va[1])
  elif va.len == 3:
    result = nimjl_call3(f, va[0], va[1], va[2])
  else:
    result = nimjl_call(f, unsafeAddr(va[0]), va.len.cint)

  nimjl_gc_pop()

## Array
# Values will need to be cast from nimjl_value to nimjl_array back and forth
proc nimjl_array_data*(values: ptr nimjl_array): pointer {.cdecl, importc.}

proc nimjl_array_dim*(a: ptr nimjl_array, dim: cint): cint {.cdecl, importc.}

proc nimjl_array_len*(a: ptr nimjl_array): cint {.cdecl, importc.}

proc nimjl_array_rank*(a: ptr nimjl_array): cint {.cdecl, importc.}

proc nimjl_new_array*(atype: ptr nimjl_value,
        dims: ptr nimjl_value): ptr nimjl_array {.cdecl, importc.}

proc nimjl_reshape_array*(atype: ptr nimjl_value, data: ptr nimjl_array,
    dims: ptr nimjl_value): ptr nimjl_array {.cdecl, importc.}

proc nimjl_ptr_to_array_1d*(atype: ptr nimjl_value, data: pointer, nel: csize_t,
    own_buffer: cint): ptr nimjl_array {.cdecl, importc.}

proc nimjl_ptr_to_array*(atype: ptr nimjl_value, data: pointer, dims: ptr nimjl_value,
    own_buffer: cint): ptr nimjl_array {.cdecl, importc.}

proc nimjl_alloc_array_1d*(atype: ptr nimjl_value,
        nr: csize_t): ptr nimjl_array {.cdecl, importc.}

proc nimjl_alloc_array_2d*(atype: ptr nimjl_value, nr: csize_t,
    nc: csize_t): ptr nimjl_array {.cdecl, importc.}

proc nimjl_alloc_array_3d*(atype: ptr nimjl_value, nr: csize_t, nc: csize_t,
    z: csize_t): ptr nimjl_array {.cdecl, importc.}

##  Array type

proc nimjl_apply_array_type_int8(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_int16(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_int32(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_int64(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_uint8(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_uint16(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_uint32(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_uint64(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_float32(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_float64(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_bool(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type_char(dim: csize_t): ptr nimjl_value {.cdecl, importc.}

proc nimjl_apply_array_type*[T](dim: int): ptr nimjl_value =
  when T is int8:
    result = nimjl_apply_array_type_int8(dim.csize_t)
  elif T is int16:
    result = nimjl_apply_array_type_int16(dim.csize_t)
  elif T is int32:
    result = nimjl_apply_array_type_int32(dim.csize_t)
  elif T is int64:
    result = nimjl_apply_array_type_int64(dim.csize_t)
  elif T is uint8:
    result = nimjl_apply_array_type_uint8(dim.csize_t)
  elif T is uint16:
    result = nimjl_apply_array_type_uint16(dim.csize_t)
  elif T is uint32:
    result = nimjl_apply_array_type_uint32(dim.csize_t)
  elif T is uint64:
    result = nimjl_apply_array_type_uint64(dim.csize_t)
  elif T is float32:
    result = nimjl_apply_array_type_float32(dim.csize_t)
  elif T is float64:
    result = nimjl_apply_array_type_float64(dim.csize_t)
  elif T is bool:
    result = nimjl_apply_array_type_bool(dim.csize_t)
  elif T is char:
    result = nimjl_apply_array_type_char(dim.csize_t)
  else:
    doAssert(false, "Type not supported")

proc nimjl_make_array*[T](data: ptr UncheckedArray[T], dims: openArray[int]): ptr nimjl_array =
  var array_type: ptr nimjl_value = nimjl_apply_array_type[T](dims.len)
  var dimStr = "("
  for d in dims:
    dimStr.add $d
    dimStr.add ","
  dimStr = dimStr & ")"
  var xDims = nimjl_eval_string(dimStr)
  nimjl_gc_push1(xDims.addr)
  result = nimjl_ptr_to_array(array_type, data, xDims, 0)
  nimjl_gc_pop()

proc nimjl_make_array*[T](data: Tensor[T]): ptr nimjl_array =
  result = nimjl_make_array(data.dataArray(), data.shape.toSeq)

proc nimjl_make_tuple*(v: tuple): ptr nimjl_value =
  var tupleStr = $v
  tupleStr = tupleStr.replace(":", "=")
  # This make tuple of a single element valid
  # (1) won't create a valid tuple -> (1,) is a valid tuple
  tupleStr = tupleStr.replace(")", ",)")
  result = nimjl_eval_string(tupleStr)

proc nimjl_make_tuple*(v: object): ptr nimjl_value =
  var tupleStr = $v
  tupleStr = tupleStr.replace(":", "=")
  # This make tuple of a single element valid
  # (1) won't create a valid tuple -> (1,) is a valid tuple
  tupleStr = tupleStr.replace(")", ",)")
  result = nimjl_eval_string(tupleStr)

File: src/csrc/nimjl.c
// #define JULIA_ENABLE_THREADING 1
#include <stdio.h>
#include <julia.h>

// JULIA_DEFINE_FAST_TLS() // only define this once, in an executable (not in a shared library) if you want fast code.

// eval
void nimjl_init() { jl_init(); }
void nimjl_gc_enable(int toggle) { jl_gc_enable(toggle); }
void nimjl_atexit_hook(int code) { jl_atexit_hook(code); }

// eval_string
jl_value_t *nimjl_eval_string(char *code) { return jl_eval_string(code); }

// Unbox func
uint8_t nimjl_unbox_uint8(jl_value_t *value) { return jl_unbox_uint8(value); }
uint16_t nimjl_unbox_uint16(jl_value_t *value) { return jl_unbox_uint16(value); }
uint32_t nimjl_unbox_uint32(jl_value_t *value) { return jl_unbox_uint32(value); }
uint64_t nimjl_unbox_uint64(jl_value_t *value) { return jl_unbox_uint64(value); }

int8_t nimjl_unbox_int8(jl_value_t *value) { return jl_unbox_int8(value); }
int16_t nimjl_unbox_int16(jl_value_t *value) { return jl_unbox_int16(value); }
int32_t nimjl_unbox_int32(jl_value_t *value) { return jl_unbox_int32(value); }
int64_t nimjl_unbox_int64(jl_value_t *value) { return jl_unbox_int64(value); }

float nimjl_unbox_float32(jl_value_t *value) { return jl_unbox_float32(value); }
double nimjl_unbox_float64(jl_value_t *value) { return jl_unbox_float64(value); }

// Box func
jl_value_t *nimjl_box_uint8(uint8_t value) { return jl_box_uint8(value); }
jl_value_t *nimjl_box_uint16(uint16_t value) { return jl_box_uint16(value); }
jl_value_t *nimjl_box_uint32(uint32_t value) { return jl_box_uint32(value); }
jl_value_t *nimjl_box_uint64(uint64_t value) { return jl_box_uint64(value); }

jl_value_t *nimjl_box_int8(int8_t value) { return jl_box_int8(value); }
jl_value_t *nimjl_box_int16(int16_t value) { return jl_box_int16(value); }
jl_value_t *nimjl_box_int32(int32_t value) { return jl_box_int32(value); }
jl_value_t *nimjl_box_int64(int64_t value) { return jl_box_int64(value); }

jl_value_t *nimjl_box_float32(float value) { return jl_box_float32(value); }
jl_value_t *nimjl_box_float64(double value) { return jl_box_float64(value); }

// Call function
jl_function_t *nimjl_get_function(jl_module_t *module, const char *name)
{
  return jl_get_function(module, name);
}

jl_value_t *nimjl_call(jl_function_t *f, jl_value_t **args, int32_t nargs)
{
  return jl_call(f, args, nargs);
}

jl_value_t *nimjl_call0(jl_function_t *f)
{
  return jl_call0(f);
}

jl_value_t *nimjl_call1(jl_function_t *f, jl_value_t *args1)
{
  return jl_call1(f, args1);
}

jl_value_t *nimjl_call2(jl_function_t *f, jl_value_t *args1, jl_value_t *args2)
{
  return jl_call2(f, args1, args2);
}

jl_value_t *nimjl_call3(jl_function_t *f, jl_value_t *args1, jl_value_t *args2, jl_value_t *args3)
{
  return jl_call3(f, args1, args2, args3);
}

// Array
void *nimjl_array_data(jl_array_t *a)
{
  return jl_array_data(a);
}

int nimjl_array_dim(jl_array_t *a, int dim)
{
  return jl_array_dim(a, dim);
}

int nimjl_array_len(jl_array_t *a)
{
  return jl_array_len(a);
}

int nimjl_array_rank(jl_array_t *a)
{
  return jl_array_rank(a);
}

jl_array_t *nimjl_new_array(jl_value_t *atype, jl_value_t *dims)
{
  return nimjl_new_array(atype, dims);
}

jl_array_t *nimjl_reshape_array(jl_value_t *atype, jl_array_t *data, jl_value_t *dims)
{
  return jl_reshape_array(atype, data, dims);
}

jl_array_t *nimjl_ptr_to_array_1d(jl_value_t *atype, void *data, size_t nel, int own_buffer)
{
  return jl_ptr_to_array_1d(atype, data, nel, own_buffer);
}

jl_array_t *nimjl_ptr_to_array(jl_value_t *atype, void *data, jl_value_t *dims, int own_buffer)
{
  return jl_ptr_to_array(atype, data, dims, own_buffer);
}

jl_array_t *nimjl_alloc_array_1d(jl_value_t *atype, size_t nr)
{
  return jl_alloc_array_1d(atype, nr);
}

jl_array_t *nimjl_alloc_array_2d(jl_value_t *atype, size_t nr, size_t nc)
{
  return jl_alloc_array_2d(atype, nr, nc);
}

jl_array_t *nimjl_alloc_array_3d(jl_value_t *atype, size_t nr, size_t nc, size_t z)
{
  return jl_alloc_array_3d(atype, nr, nc, z);
}
// Array type
jl_value_t *nimjl_apply_array_type_int8(size_t dim)
{
  return jl_apply_array_type(jl_int8_type, dim);
}

jl_value_t *nimjl_apply_array_type_int16(size_t dim)
{
  return jl_apply_array_type(jl_int16_type, dim);
}

jl_value_t *nimjl_apply_array_type_int32(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_int32_type, dim);
}

jl_value_t *nimjl_apply_array_type_int64(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_int64_type, dim);
}

jl_value_t *nimjl_apply_array_type_uint8(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_uint8_type, dim);
}

jl_value_t *nimjl_apply_array_type_uint16(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_uint16_type, dim);
}

jl_value_t *nimjl_apply_array_type_uint32(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_uint32_type, dim);
}

jl_value_t *nimjl_apply_array_type_uint64(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_uint64_type, dim);
}

jl_value_t *nimjl_apply_array_type_float32(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_float32_type, dim);
}

jl_value_t *nimjl_apply_array_type_float64(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_float64_type, dim);
}

jl_value_t *nimjl_apply_array_type_bool(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_bool_type, dim);
}

jl_value_t *nimjl_apply_array_type_char(size_t dim)
{
  return jl_apply_array_type((jl_value_t *)jl_char_type, dim);
}

inline void nimjl_gc_push1(void *a)
{
  JL_GC_PUSH1(a);
}

inline void nimjl_gc_push2(void *a, void *b)
{
  JL_GC_PUSH2(a, b);
}

inline void nimjl_gc_push3(void *a, void *b, void *c)
{
  JL_GC_PUSH3(a, b, c);
}

inline void nimjl_gc_push4(void *a, void *b, void *c, void *d)
{
  JL_GC_PUSH4(a, b, c, d);
}

inline void nimjl_gc_push5(void *a, void *b, void *c, void *d, void *e)
{
  JL_GC_PUSH5(a, b, c, d, e);
}

inline void nimjl_gc_push6(void *a, void *b, void *c, void *d, void *e, void *f)
{
  JL_GC_PUSH6(a, b, c, d, e, f);
}

inline void nimjl_gc_pushargs(jl_value_t **a, size_t n)
{
  JL_GC_PUSHARGS(a, n);
}

inline void nimjl_gc_pop()
{
  JL_GC_POP();
}

jl_value_t *nimjl_exception_occurred()
{
  return jl_exception_occurred();
}

char *nimjl_typeof_str(jl_value_t *v)
{
  return jl_typeof_str(v);
}

char *nimjl_string_ptr(jl_value_t *v)
{
  return jl_string_ptr(v);
}
